File: zad1/ClientApp/src/models/selection/TournamentSelection.ts
import ISelection from "./ISelection";
import Population from "../population/Population";
import Individual from "../Individual";
import _ from "lodash";

export default class TournamentSelection implements ISelection {
    constructor(public n: number) {}

    evaluate(population: Population): Individual {
        if (this.n > population.individuals.length)
            this.n = population.individuals.length;

        // const randomIndividuals = new Array<Individual>(this.n)
        //     .fill(null)
        //     .map(
        //         () => population.individuals[~~(Math.random() * this.n)]
        //     );

        // const distances = randomIndividuals.map(individual =>
        //     individual.calculateTotalDistance()
        // );

        // const bestDistance = Math.min(...distances);
        // const bestIndividualIndex = distances.findIndex(
        //     distance => bestDistance === distance
        // );

        // return _.cloneDeep(randomIndividuals[bestIndividualIndex]);

        const chosenIndividualIndices = [];
        for (let i = 0; i < this.n; i++)
            chosenIndividualIndices.push(
                ~~(Math.random() * population.individuals.length)
            );

        let chosenIndex = chosenIndividualIndices[0];
        let individual = population.individuals[chosenIndex];
        let minDistance = individual.calculateTotalDistance();

        for (let i = 0; i < chosenIndividualIndices.length; i++) {
            let individualIndex = chosenIndividualIndices[i];
            individual = population.individuals[individualIndex];
            const distance = individual.calculateTotalDistance();

            if (distance < minDistance) {
                chosenIndex = individualIndex;
                minDistance = distance;
            }
        }

        return _.cloneDeep(population.individuals[chosenIndex]);
    }
}

File: zad1/Algorithms/Selection/TournamentSelection.cs
using System;
using System.Collections.Generic;
using back.Models;
using back.Models.Population;

namespace zad1.Algorithms.Selection
{
    public class TournamentSelection : ISelection
    {
        public int N { get; }

        public TournamentSelection(int n)
        {
            N = n;
        }

        public Individual Evaluate(Population population)
        {
            var chosenIndividualIndices = new List<int>();
            var rng = new Random();

            for (var i = 0; i < N; i++)
                chosenIndividualIndices.Add(rng.Next(population.Individuals.Count));

            var chosenIndex = chosenIndividualIndices[0];
            var individual = population.Individuals[chosenIndex];
            var minDistance = individual.CalculateTotalDistance();

            foreach (var individualIndex in chosenIndividualIndices)
            {
                individual = population.Individuals[individualIndex];
                var distance = individual.CalculateTotalDistance();

                if (distance < minDistance)
                {
                    chosenIndex = individualIndex;
                    minDistance = distance;
                }
            }

            return (Individual) population.Individuals[chosenIndex].Clone();
        }
    }
}